File: baseclient/java/src/main/java/com/aliyun/ccp/baseclient/BaseClient.java
package com.aliyun.ccp.baseclient;

import com.aliyun.ccp.baseclient.credentials.AccessTokenCredential;
import com.aliyun.credentials.AccessKeyCredential;
import com.aliyun.credentials.StsCredential;
import com.aliyun.tea.TeaRequest;
import com.aliyun.tea.TeaResponse;
import com.aliyun.tea.utils.StringUtils;
import com.google.gson.Gson;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.Map.Entry;

public class BaseClient {
    public final static String URL_ENCODING = "UTF-8";
    private static final String ALGORITHM_NAME = "HmacSHA1";
    protected String _domainId;
    protected String _protocol;
    protected Object _endpoint;
    protected Map<String, Object> _config;
    protected AccessKeyCredential _accessKeyCredential;
    protected AccessTokenCredential _accessTokenCredential;
    protected StsCredential _stsCredential;

    public BaseClient(Map<String, Object> config) {
        this._domainId = (String) config.get("domainId");
        this._protocol = (String) config.get("protocol");
        this._endpoint = (String) config.get("endpoint");
        setCredential(config);
        this._config = config;
    }

    private void setCredential(Map<String, Object> config) {
        if (!StringUtils.isEmpty(config.get("securityToken"))) {
            _stsCredential = new StsCredential((String) config.get("accessKeyId"),
                    (String) config.get("accessKeySecret"), (String) config.get("securityToken"));
        } else if (!StringUtils.isEmpty(config.get("accessToken"))) {
            _accessTokenCredential = new AccessTokenCredential(config);
        } else if (!StringUtils.isEmpty(config.get("accessKeyId"))) {
            _accessKeyCredential = new AccessKeyCredential((String) config.get("accessKeyId"),
                    (String) config.get("accessKeySecret"));
        }
    }

    public String _getRFC2616Date() {
        SimpleDateFormat df = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US);
        df.setTimeZone(TimeZone.getTimeZone("GMT"));
        return df.format(new Date());
    }

    public String _getUserAgent() {
        return "";
    }

    public String getCanonicalizedHeaders(Map<String, String> headers) {
        String prefix = "x-acs-";
        Set<String> keys = headers.keySet();
        List<String> canonicalizedKeys = new ArrayList<>();
        for (String key : keys) {
            if (key.startsWith(prefix)) {
                canonicalizedKeys.add(key);
            }
        }
        String[] canonicalizedKeysArray = canonicalizedKeys.toArray(new String[canonicalizedKeys.size()]);
        Arrays.sort(canonicalizedKeysArray);
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < canonicalizedKeysArray.length; i++) {
            if (i > 0) {
                result.append("\n");
            }
            String key = canonicalizedKeysArray[i];
            result.append(key);
            result.append(":");
            result.append(headers.get(key).trim());
        }
        return result.toString();
    }

    public String getCanonicalizedResource(String pathname, Map<String, String> query) {
        String[] keys = query.keySet().toArray(new String[query.size()]);
        if (keys.length <= 0) {
            return pathname;
        }
        StringBuilder result = new StringBuilder(pathname);
        result.append("?");
        String key;
        for (int i = 0; i < keys.length; i++) {
            key = keys[i];
            result.append(key);
            result.append("=");
            result.append(query.get(key));
        }
        return result.toString();
    }


    public String _getSignature(TeaRequest request) throws UnsupportedEncodingException, NoSuchAlgorithmException, InvalidKeyException {
        String method = request.method;
        String pathname = request.pathname;
        if (null != this._getSecurityToken()) {
            request.headers.put("x-acs-security-token", this._getSecurityToken());
        }
        Map<String, String> headers = request.headers;
        Map<String, String> query = request.query;
        String accept = headers.get("accept") == null ? "" : headers.get("accept");
        String contentMD5 = headers.get("content-md5") == null ? "" : headers.get("content-md5");
        String contentType = headers.get("content-type") == null ? "" : headers.get("content-type");
        String date = headers.get("date") == null ? "" : headers.get("date");
        String header = method + "\n" + accept + "\n" + contentMD5 + "\n" + contentType + "\n" + date;
        String canonicalizedHeaders = getCanonicalizedHeaders(headers);
        String canonicalizedResource = getCanonicalizedResource(pathname, query);
        String stringToSign = header + "\n" + canonicalizedHeaders + "\n" + canonicalizedResource;
        Mac mac = Mac.getInstance(ALGORITHM_NAME);
        mac.init(new SecretKeySpec(((String) _config.get("accessKeySecret")).getBytes(URL_ENCODING), ALGORITHM_NAME));
        byte[] signData = mac.doFinal(stringToSign.getBytes(URL_ENCODING));
        return Base64.getEncoder().encodeToString(signData);
    }

    public String percentEncode(String value) throws UnsupportedEncodingException {
        return value != null
                ? URLEncoder.encode(value, URL_ENCODING).replace("+", "%20").replace("*", "%2A").replace("%7E", "~")
                : null;
    }

    protected String _getProtocol(String protocol, String https) {
        if (_notEmpty(protocol)) {
            return protocol;
        }
        return https;
    }

    protected Object _getHost(Object endpoint, String s) {
        if (_notEmpty(String.valueOf(endpoint))) {
            return endpoint;
        }
        return s;
    }

    protected boolean _notEmpty(String str) {
        if (null == str || "null".equals(str.trim())) {
            return false;
        }
        if (str.trim().length() <= 0) {
            return false;
        }
        return true;
    }

    protected String _toJSONString(Object request) {
        return new Gson().toJson(request);
    }

    protected boolean _isStatusCode(TeaResponse response_, int code) {
        return response_.statusCode == code;
    }

    @SuppressWarnings("unchecked")
    protected Map<String, Object> _readAsJSON(TeaResponse response_) throws IOException {
        String body = response_.getResponseBody();
        Gson gson = new Gson();
        Map<String, Object> map = gson.fromJson(body, Map.class);
        return map;
    }

    public Map<String, String> _toQuery(Map<String, Object> request) {
        Map<String, String> map = new HashMap<String, String>();
        for (Entry<String, Object> entry : request.entrySet()) {
            map.put(entry.getKey(), entry.getValue().toString());
        }
        return map;
    }

    public int _defaultNumber(Integer maxAttempts, int defaultNumber) {
        if (maxAttempts != null && maxAttempts >= 0) {
            return maxAttempts;
        }
        return defaultNumber;
    }

    public String _default(Object maxAttempts, String defaultStr) {
        if (null != maxAttempts) {
            return String.valueOf(maxAttempts);
        }
        return defaultStr;
    }

    public String _getAccessKeyId() {
        if (null != this._accessKeyCredential) {
            return this._accessKeyCredential.getAccessKeyId();
        } else if (null != this._stsCredential) {
            return this._stsCredential.getAccessKeyId();
        }
        return null;
    }

    public String _getAccessKeySecret() {
        if (null != this._accessKeyCredential) {
            return this._accessKeyCredential.getAccessKeySecret();
        } else if (null != this._stsCredential) {
            return this._stsCredential.getAccessKeySecret();
        }
        return null;
    }

    public void _setAccessKeyId(String accessKeyId) {
        if (null != this._stsCredential) {
            this._stsCredential.setAccessKeyId(accessKeyId);
        }
    }

    public void _setAccessKeySecret(String accessKeySecret) {
        if (null != this._stsCredential) {
            this._stsCredential.setAccessKeySecret(accessKeySecret);
        }
    }

    public String _getSecurityToken() {
        if (null != this._stsCredential) {
            return this._stsCredential.getSecurityToken();
        }
        return null;
    }

    public void _setSecurityToken(String securityToken) {
        if (null != this._stsCredential) {
            this._stsCredential.setSecurityToken(securityToken);
        }
    }

    public String _getAccessToken() throws Exception {
        if (null != _accessTokenCredential) {
            return this._accessTokenCredential.getAccessToken();
        }
        return null;
    }

    public void _setAccessToken(String accessToken) {
        if (null != _accessTokenCredential) {
            this._accessTokenCredential.setAccessToken(accessToken);
        }
    }

    public void _setRefreshToken(String refreshToken) {
        if (null != _accessTokenCredential) {
            this._accessTokenCredential.setRefreshToken(refreshToken);
        }
    }

    public void _setExpireTime(String expireTime) {
        if (null != _accessTokenCredential) {
            this._accessTokenCredential.setExpireTime(expireTime);
        }
    }

    public String _getRefreshToken() {
        if (null != _accessTokenCredential) {
            return this._accessTokenCredential.getRefreshToken();
        }
        return null;
    }

    public String _getExpireTime() {
        if (null != _accessTokenCredential) {
            return this._accessTokenCredential.getExpireTime();
        }
        return null;
    }
}

File: ccppath-sdk/swift/Sources/AliyunCCPSDK/BaseClient.swift
//
// Created by Axios on 2020/2/19.
//

import Foundation
import Tea
import CryptoSwift

open class BaseClient {
    var sdkVersion = "0.0.1";

    var accessKeyId: String = "";
    var accessKeySecret: String = "";
    var accessToken: String = "";
    var _userAgent: String = "";

    var _protocol: String = "https";
    var _endpoint: String = "";
    var _domainId: String = "";
    var _nickname: String = "";
    var config: [String: Any];

    static var _defaultUserAgent: String = "";

    init(_ config: [String: Any]) {
        self._domainId = config["domainId"] as! String;
        self._protocol = config["protocol_"] as! String;
        self._endpoint = config["endpoint"] as! String;
        self._nickname = config["nickname"] as! String;
        self.accessKeyId = config["accessKeyId"] as! String;
        self.accessKeySecret = config["accessKeySecret"] as! String;
        self.config = config;
    }

    private func _getDefaultUserAgent() -> String {
        if BaseClient._defaultUserAgent.count > 0 {
            return BaseClient._defaultUserAgent;
        }
        let os = ProcessInfo().operatingSystemVersion;
        let OSVersion: String = String(os.majorVersion) + "." + String(os.minorVersion) + "." + String(os.patchVersion);
        BaseClient._defaultUserAgent = "Alibaba Cloud (" + OSVersion + ") ";
        BaseClient._defaultUserAgent += " Tea-iOS-SDK/" + self.sdkVersion;
        return BaseClient._defaultUserAgent;
    }

    public func _getPathname(_ nickname: String, _ path: String) -> String {
        if (nickname.isEmpty) {
            return path;
        }
        return "/" + nickname + path;
    }

    public func _getUserAgent() -> String {
        if !self._userAgent.isEmpty {
            return self._getDefaultUserAgent() + " " + self._userAgent;
        }
        return self._getDefaultUserAgent();
    }

    public func _default(_ str: String, _ defaultStr: String) -> String {
        if str.isEmpty {
            return defaultStr;
        }
        return str;
    }

    public func _notEmpty(_ s: Any?) -> Bool {
        if s == nil {
            return false
        } else if s as? Int == 0 {
            return false
        }
        let str: String = s as! String;
        return str.count > 0 ? true : false;
    }

    public func _toJSONString(_ dict: [TeaModel]) -> String {
        var dic: [[String: Any]] = [[String: Any]]();
        for model in dict {
            dic.append(model.toMap());
        }
        guard let data = try? JSONSerialization.data(withJSONObject: dic, options: []) else {
            return "";
        }
        let json = String(data: data, encoding: String.Encoding.utf8)!;
        return json;
    }

    public func _defaultNumber(_ number: CLong, _ defaultNumber: CLong) -> CLong {
        if number > 0 {
            return number;
        }
        return defaultNumber;
    }

    public func _readAsJSON(_ response: TeaResponse) -> [String: AnyObject] {
        guard let data = try? JSONSerialization.jsonObject(with: response.data, options: .mutableContainers) as? [String: AnyObject] else {
            return [String: AnyObject]();
        }
        return data
    }

    public func _getAccessKeyId() -> String {
        return self.accessKeyId;
    }

    public func _getAccessKeySecret() -> String {
        return self.accessKeySecret;
    }

    public func _getAccessToken() -> String {
        return self.accessToken;
    }

    public func _getProtocol(_ protocol: String, _ defaultProtocol: String) -> String {
        if self._protocol.isEmpty {
            return defaultProtocol;
        }
        return self._protocol;
    }

    public func _getRFC2616Date() -> String {
        let curr = Date();
        let formatter = DateFormatter();
        formatter.locale = Locale(identifier: "en_US");
        formatter.timeZone = TimeZone(identifier: "GMT");
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss";
        return formatter.string(from: curr) + " GMT";
    }

    public func _getHost(_ endpoint: String, _ domain: String) -> String {
        if self._endpoint.isEmpty {
            return domain;
        }
        return self._endpoint;
    }

    public func _getSignature(_ request: TeaRequest) -> String {
        let stringToSign: String = self._getStringToSign(request)
        let r = try! HMAC(key: self.accessKeySecret, variant: .sha1).authenticate(stringToSign.bytes)
        return r.toBase64() ?? "";
    }

    public func _getStringToSign(_ request: TeaRequest) -> String {
        let method: String = request.method;
        let pathname: String = request.pathname;
        let headers: [String: String] = request.headers;
        let query: [String: Any] = request.query;

        let accept = headers["accept"] ?? "";
        let contentMD5 = headers["content-md5"] ?? "";
        let contentType = headers["content-type"] ?? "";
        let date = headers["date"] ?? "";

        let headerStr = String(method) + "\n" + String(accept) + "\n" + String(contentMD5) + "\n" + String(contentType) + "\n" + String(date);
        let canonicalizedHeaders: String = self.getCanonicalizedHeaders(headers: headers);
        let canonicalizedResource: String = getCanonicalizedResource(pathname: pathname, query: query);
        return headerStr + "\n" + canonicalizedHeaders + "\n" + canonicalizedResource;
    }

    public func getCanonicalizedHeaders(headers: [String: String]) -> String {
        let prefix: String = "x-acs-";
        var canonicalizedKeys: [String] = [String]();
        for (key, _) in headers {
            if (key.hasPrefix(prefix)) {
                canonicalizedKeys.append(key);
            }
        }
        canonicalizedKeys.sort();
        var result: String = "";
        var n = 0;
        for key in canonicalizedKeys {
            if n != 0 {
                result.append(contentsOf: "\n");
            }
            result.append(contentsOf: key);
            result.append(contentsOf: ":");
            result.append(contentsOf: headers[key]?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "");
            n += 1;
        }
        return result;
    }

    public func getCanonicalizedResource(pathname: String, query: [String: Any]) -> String {
        if (query.count <= 0) {
            return pathname;
        }
        var keys: [String] = [String]();
        for (key, _) in query {
            keys.append(key);
        }
        keys.sort();
        var result: String = pathname + "?";
        for key in keys {
            result.append(key);
            result.append("=");
            result.append(query[key] as! String);
        }
        return result;
    }

    public func _isStatusCode(_ response: TeaResponse?, _ status: Int) -> Bool {
        return Int(response?.statusCode ?? "") == status;
    }
}